pep-0xxx.rst

```diff
@@ -0,0 +1,727 @@
+            during the TLS handshake. This may be advertised using either or
+            both of ALPN or NPN.
+
+            ``protocols`` should be a list of acceptable protocols in the form
+            of ``NextProtocol`` objects, such as ``[H2, HTTP1]``, ordered by
+            preference. The selection of the protocol will happen during the
+            handshake, and will use whatever protocol negotiation mechanisms
+            are available and supported by both peers.
+
+            If the TLS implementation doesn't support protocol negotiation,
+            this method will raise ``NotImplementedError``.
+            """
+
+        @abstractmethod
+        def set_sni_callback(self, callback: Optional[ServerNameCallback]) -> None:
+            """
+            Register a callback function that will be called after the TLS
+            Client Hello handshake message has been received by the TLS server
+            when the TLS client specifies a server name indication.
+
+            Only one callback can be set per ``Context``. If ``callback`` is
+            ``None`` then the callback is disabled. Calling this function a
+            subsequent time will disable the previously registered callback.
+
+            The ``callback`` function will be called with three arguments: the
+            first will be the ``TLSBufferObject`` for the connection; the
+            second will be a string that represents the server name that the
+            client is intending to communicate (or ``None`` if the TLS Client
+            Hello does not contain a server name); and the third argument will
+            be the original ``Context``. The server name argument will be the
+            IDNA *decoded* server name.
+
+            The ``callback`` must return ``None`` to allow negotiation to
+            continue. Other return values signal errors. Attempting to control
+            what error is signaled by the underlying TLS implementation is not
+            specified in this API, but is up to the concrete implementation to
+            handle.
+            """
+
+        @abstractmethod
+        def set_version_range(self, lower_bound=None: Optional[TLSVersion],
+                              upper_bound=None: Optional[TLSVersion]) -> None:
+            """
+            Set the minumum and maximum versions of TLS that should be allowed
+            on TLS connections made by this context.
+
+            If present, ``lower_bound`` will set the lowest acceptable TLS
+            version. If present, ``upper_bound`` will set the highest
+            acceptable TLS version. If either argument is ``None``, this will
+            leave that bound unchanged.
+            """
+
+        @abstractmethod
+        def wrap_socket(self, socket: socket.socket, server_side=False: bool,
+                        auto_handshake=True: bool,
+                        server_hostname=None: Optional[str]) -> TLSWrappedSocket:
```

Comment:
Well, the `suppress_ragged_eofs` thing is clearly not a necessary feature. But in stdlib `ssl`, the default is to ignore unclean shutdowns (`suppress_ragged_eofs=True`), which is a security problem in some cases. So we should *at the least* provide an option to act like `suppress_ragged_eofs=False`. The other option is to *only* act like `suppress_ragged_eofs=False`. This means everyone implementing common protocols like HTTPS will need to catch and ignore the ragged EOF exceptions, and complicate porting from stdlib `ssl` (people are probably depending on `suppress_ragged_eofs=True` without realizing it). So... either an option or only-`False` seem reasonable to me, but either way I think the spec needs to explicitly say what it's doing :-).

Regarding `session`: I thought it needs to be exposed because e.g. servers may need to stick the session ticket in a database? I guess the alternative you're proposing is that each TLS implementation would hard-code some in-memory LRU ticket cache, maybe with some extension to tune the cache size etc.? This is beyond my TLS knowledge depth, just wanted to make sure the issue didn't get missed :-)